### PR TITLE
Support Config.noguestsecurity

### PR DESCRIPTION
**Pull request by Zarel**
This Config option, intended purely for development test servers, allows users to log in without a signed token, by using `/trn [USERNAME]`. This allows logins without a login server.

The rest of the server has been changed to make sure unregistered users never have any powers they shouldn't have.

### CODE REVIEW

**Comment by Zarel** [comment]
@monsanto @xfix I'd very much like some assurance that I'm not accidentally introducing some sort of vulnerability.

I want this to be safe to enable on development servers without actually letting malicious users take over your server (DoSing is fine).

File: server/user-groups.ts [hunk 1]
```diff
@@ -213,11 +213,14 @@ export class RoomAuth extends Auth {
 		super();
 		this.room = room;
 	}
-	get(user: ID | User): GroupSymbol {
+	get(userOrID: ID | User): GroupSymbol {
+		const user = typeof userOrID === 'string' ? null : userOrID as User;
+		const id = user ? user.id : userOrID as ID;
+		if (user && !user.registered) return Auth.defaultSymbol();
```
Comment (xfix) [inline server/user-groups.ts:219]: This means that unregistered users won't get player rank in battles, is this intended?

**Comment by Zarel** [comment]
Refs #7618

**Comment by monsanto** [comment]
We need to be extra sure to not do permission checks with usernames in the server, or at least ensure that we are guarding the check with another check that the username is registered. I know the private PM log code has username checks, if that was to ever be upstreamed. I think the friends list does too. I think the rest of the username checks are in the client/website and are gated by the login server.

Besides that & Annika's concern seems OK to me

**Comment by mia-pi-git** [comment]
> I think the friends list does too.

The friends list is gated on being autoconfirmed.

File: server/user-groups.ts (hunk 1 above)
Comment (Zarel) [reply server/user-groups.ts:219]: Ah, no, that's not intended. But it may be preferable?

**Comment by Zarel** [comment]
> Testing this locally, if I logged in as Annika with authentication, then in a private browsing window did `/trn Annika` I'd have access to the same user object with console access and so forth. Also, even if I didn't already have a session, I would gain trusted access, which is not ideal.

I thought we had code to prevent unregistered users from merging? I guess you bypassed because you had the same IP? I can add a guard for that.

File: server/user-groups.ts (hunk 1 above)
Comment (Zarel) [reply server/user-groups.ts:219]: We could hardcode an exception for temporary rooms.

**Comment by Zarel** [comment]
One thing I mulled was to just not allow people to take names of trusted users. I decided against that because I thought it might interfere with poke-env's use-case. But maybe I should revisit that post.

@hsahovic would that work? Only allow empty tokens for users that don't have any authority on the server?

**Comment by hsahovic** [comment]
In principle, I think so. What would count as a user with authority on the server?

**Comment by Zarel** [comment]
> In principle, I think so. What would count as a user with authority on the server?

"Known" means "any roomauth at all". (Which includes if a user has battled within the last ten minutes, no one else can use that account.)

There's also the weaker "trusted" which means "global voice or higher, or roomdriver or higher in a public/hidden room".

**Comment by hsahovic** [comment]
In that case, I think that this would be too restrictive - I can think of several use cases which would require authenticating the same user several times in a couple of minutes. Could such a limitation be avoided?

**Comment by Zarel** [comment]
It sounds like the "trusted" line would work for you, then?

**Comment by monsanto** [comment]
I'd prefer "trusted" if possible, but if that is too restrictive, even just disallowing sysops would be a nice defense-in-depth thing to do

**Comment by hsahovic** [comment]
Yeah, it sounds like "trusted" would work. I'm a bit shaky on what global voice / roomdriver entails though.

**Comment by Zarel** [comment]
Are you familiar with PS's rank system?

`+` Voice (general recognition with no moderation powers)
`%` Driver (junior moderator)
`@` Moderator
`&` Administrator

They can be global ranks or room ranks.

**Comment by hsahovic** [comment]
Do new users have global voice?

**Comment by AnnikaCodes** [comment]
They do not.

**Comment by hsahovic** [comment]
In this case "Trusted" shoud be fine :)

File: server/users.ts [hunk 2]
```diff
@@ -587,28 +587,18 @@ export class User extends Chat.MessageContext {
...
-					console.log(`desynced roomgame ${roomid} renaming ${this.id} -> ${userid}`);
-					continue;
-				}
-				if (room.game.allowRenames || !this.named) continue;
-				this.popup(`You can't change your name right now because you're in ${room.game.title}, which doesn't allow renaming.`);
-				return false;
+	async validateToken(token: string, name: string, userid: ID, connection: Connection) {
+		if (!token && Config.noguestsecurity) {
+			if (Users.isTrusted(userid)) {
+				this.send(`|nametaken|${name}|You need an authentication token to be a trusted user.`);
```
Comment (AnnikaCodes) [inline server/users.ts:593]:
```suggestion
				this.send(`|nametaken|${name}|You need an authentication token to log in as a trusted user.`);
```

Comment (AnnikaCodes) [reply server/users.ts:593]: The original error messages reads to me as "you can log in, you just won't be marked as trusted"

File: server/users.ts [hunk 3]
```diff
@@ -730,27 +745,36 @@ export class User extends Chat.MessageContext {
...
+			// unregistered users can only merge in limited situations
+			let canMerge = registered && conflictUser.registered;
+			if (
+				!registered && !conflictUser.registered && conflictUser.latestIp !== this.latestIp &&
+				!conflictUser.connected
+			) {
+				canMerge = true;
+			}
+			if (!canMerge) {
+				if (newlyRegistered && registered) {
```
Comment (Zarel) [inline server/users.ts:761]:
```suggestion
				if (registered && !conflictUser.registered) {
```

File: server/users.ts [hunk 4]
```diff
@@ -730,27 +745,36 @@ export class User extends Chat.MessageContext {
...
-			if (newlyRegistered && userType !== '1') {
-				if (conflictUser !== this) conflictUser.resetName();
-			} else {
-				this.send(`|nametaken|${name}|Someone is already using the name "${conflictUser.name}".`);
-				return false;
+		if (conflictUser) {
+			// unregistered users can only merge in limited situations
+			let canMerge = registered && conflictUser.registered;
+			if (
+				!registered && !conflictUser.registered && conflictUser.latestIp !== this.latestIp &&
```
Comment (Zarel) [inline server/users.ts:755]:
```suggestion
				!registered && !conflictUser.registered && conflictUser.latestIp === this.latestIp &&
```

File: config/config-example.js [hunk 5]
```diff
@@ -390,6 +390,16 @@ exports.inactiveuserthreshold = 1000 * 60 * 60;
  */
 exports.autolockdown = true;
 
+/**
+ * noguestsecurity - purely for development servers: allows logging in without
+ * a signed token: simply send `/trn [USERNAME]`. This allows using PS without
+ * a login server.
+ *
+ * Logging in this way will make you considered an unregistered user and grant
+ * no authority.
```
Comment (Zarel) [inline config/config-example.js:399]:
```suggestion
 * Logging in this way will make you considered an unregistered user and grant
 * no authority. You cannot log into a trusted (g+/r%) user account this way.
```

**Comment by Zarel** [comment]
Safe to hotpatch! Won't take effect upon a restart, though.

File: server/user-groups.ts (hunk 1 above)
Comment (Zarel) [reply server/user-groups.ts]: (I ended up replacing this with a trusted user check, which no longer has this problem.)

**Comment by Zarel** [comment]
3/3!

**Pull request merged**